runtime/near-evm-runner/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+        let value =
+            if self.attached_deposit == 0 { None } else { Some(U256::from(self.attached_deposit)) };
+        interpreter::call(self, &origin, &sender, value, 0, &contract_address, &input, true)
+            .map(|rd| rd.to_vec())
+    }
+
+    /// Make an EVM call via a meta transaction pattern.
+    /// Specifically, providing signature and NEAREvm message that determines which contract and arguments to be called.
+    /// Format
+    /// 0..95: signature: v - 32 bytes, s - 32 bytes, r - 32 bytes
+    /// 96..115: contract_id: address for contract to call
+    /// 116..: RLP encoded arguments.
+    pub fn meta_call_function(&mut self, args: Vec<u8>) -> Result<Vec<u8>> {
+        if args.len() <= 148 {
+            return Err(VMLogicError::EvmError(EvmError::ArgumentParseError));
+        }
+        let mut signature: [u8; 96] = [0; 96];
+        signature.copy_from_slice(&args[..96]);
+        let nonce = U256::from_big_endian(&args[96..128]);
+        let args = &args[128..];
+        let sender = ecrecover_address(
+            &prepare_meta_call_args(&self.domain_separator, &self.account_id, nonce, args),
+            &signature,
+        );
+        if sender == Address::zero() {
+            return Err(VMLogicError::EvmError(EvmError::InvalidEcRecoverSignature));
+        }
+        if self.next_nonce(&sender)? != nonce {
+            return Err(VMLogicError::EvmError(EvmError::InvalidNonce));
+        }
+        let contract_address = Address::from_slice(&args[..20]);
+        let input = &args[20..];
+        self.add_balance(&sender, U256::from(self.attached_deposit))?;
+        let value =
+            if self.attached_deposit == 0 { None } else { Some(U256::from(self.attached_deposit)) };
+        interpreter::call(self, &sender, &sender, value, 0, &contract_address, &input, true)
+            .map(|rd| rd.to_vec())
+    }
+
+    /// Make an EVM transaction. Calls `contract_address` with `encoded_input`. Execution
+    /// continues until all EVM messages have been processed. We expect this to behave identically
+    /// to an Ethereum transaction, however there may be some edge cases.
+    ///
+    /// This function serves the eth_call functionality, and will NOT apply state changes.
+    pub fn view_call_function(&mut self, args: Vec<u8>) -> Result<Vec<u8>> {
+        let args = ViewCallArgs::try_from_slice(&args)
+            .map_err(|_| VMLogicError::EvmError(EvmError::ArgumentParseError))?;
+        let sender = Address::from(&args.sender);
+        let attached_amount = U256::from(args.amount);
+        self.add_balance(&sender, attached_amount)?;
+        let result = interpreter::call(
+            self,
+            &sender,
+            &sender,
+            Some(attached_amount),
+            0,
+            &Address::from(&args.address),
+            &args.args,
+            false,
+        )
+        .map(|rd| rd.to_vec());
+        self.sub_balance(&sender, attached_amount)?;
```

Comment:
Wondering why we have `sub_balance` here but not in `pub fn call_function(` a little ways up?

Reply:
because view function doesn't actually change the balance - but for the call itself wants to emulate to have the balance.

Given view call can be called within a transaction - hence state will be changing if we just `add_balance` - want to subtract it out.